src/Nethermind/Nethermind.Core/Crypto/KeccakCache.cs

```diff
@@ -0,0 +1,226 @@
+/// </summary>
+public static unsafe class KeccakCache
+{
+    /// <summary>
+    /// Count is defined as a +1 over bucket mask. In the future, just change the mask as the main parameter.
+    /// </summary>
+    public const int Count = BucketMask + 1;
+    private const int BucketMask = 0x0000_FFFF;
+
+    private static readonly Entry* Memory;
+
+    static KeccakCache()
+    {
+        const UIntPtr size = Count * Entry.Size;
+
+        // Aligned, so that no torn reads if fields of Entry are properly aligned.
+        Memory = (Entry*)NativeMemory.AlignedAlloc(size, Entry.Size);
+        NativeMemory.Clear(Memory, size);
+    }
+
+    [SkipLocalsInit]
+    public static ValueHash256 Compute(ReadOnlySpan<byte> input)
+    {
+        ComputeTo(input, out ValueHash256 keccak256);
+        return keccak256;
+    }
+
+    [SkipLocalsInit]
+    public static void ComputeTo(ReadOnlySpan<byte> input, out ValueHash256 keccak256)
+    {
+        // Special cases jump forward as unpredicted
+        if (input.Length == 0 || input.Length > Entry.MaxPayloadLength)
+        {
+            goto Uncommon;
+        }
+
+        int hashCode = input.FastHash();
+        uint index = (uint)hashCode & BucketMask;
+
+        Debug.Assert(index < Count);
+
+        ref Entry e = ref Unsafe.Add(ref Unsafe.AsRef<Entry>(Memory), index);
+
+        // Read aligned, volatile, won't be torn, check with computed
+        if (Volatile.Read(ref e.HashCode) == hashCode)
+        {
+            // There's a possibility of a hit, try lock.
+            int lockAndLength = Volatile.Read(ref e.LockAndLength);
+            // Lock by negating length if the length is the same size as input.
+            if (lockAndLength == input.Length && Interlocked.CompareExchange(ref e.LockAndLength, -lockAndLength, lockAndLength) == lockAndLength)
+            {
+                if (e.HashCode != hashCode)
+                {
+                    // The value has been changed between reading and taking a lock.
+                    // Release the lock and compute.
+                    Volatile.Write(ref e.LockAndLength, lockAndLength);
+                    goto Compute;
+                }
+
+                // Take local copy of the payload and hash, to release the lock as soon as possible and make a key comparison without holding it.
+                // Local copy of 8+16+64 payload bytes.
+                Payload copy = e.Value;
+                // Copy Keccak256 directly to the local return variable, since we will overwrite if no match anyway.
+                keccak256 = e.Keccak256;
+
+                // Release the lock, by setting back to unnegated length.
+                Volatile.Write(ref e.LockAndLength, lockAndLength);
+
+                // Lengths are equal, the input length can be used without any additional operation.
+                if (input.Length == 32)
```

Comment:
Is it faster this way? My benchmarks showed that the equality wasn't the big deal here. Especially as it falls almost right into the optimized path of vectorized?